sql/core/src/main/scala/org/apache/spark/sql/execution/python/BatchEvalPythonExec.scala

```diff
@@ -17,18 +17,270 @@
 
 package org.apache.spark.sql.execution.python
 
+import java.io._
+
 import scala.collection.JavaConverters._
 import scala.collection.mutable.ArrayBuffer
 
 import net.razorvine.pickle.{Pickler, Unpickler}
 
-import org.apache.spark.TaskContext
+import org.apache.spark.{SparkEnv, TaskContext}
 import org.apache.spark.api.python.{ChainedPythonFunctions, PythonRunner}
+import org.apache.spark.memory.{MemoryConsumer, TaskMemoryManager}
 import org.apache.spark.rdd.RDD
 import org.apache.spark.sql.catalyst.InternalRow
 import org.apache.spark.sql.catalyst.expressions._
 import org.apache.spark.sql.execution.SparkPlan
 import org.apache.spark.sql.types.{DataType, StructField, StructType}
+import org.apache.spark.unsafe.Platform
+import org.apache.spark.unsafe.memory.MemoryBlock
+import org.apache.spark.util.{CompletionIterator, Utils}
+
+
+/**
+ * A RowQueue is an FIFO queue for UnsafeRow.
+ */
+private[python] trait RowQueue {
+  /**
+   * Add a row to the end of it, returns true iff the row has added into it.
+   */
+  def add(row: UnsafeRow): Boolean
+
+  /**
+   * Retrieve and remove the first row, returns null if it's empty.
+   */
+  def remove(): UnsafeRow
+
+  /**
+   * Cleanup all the resources.
+   */
+  def close(): Unit
+}
+
+/**
+ * A RowQueue that is based on in-memory page. UnsafeRows are appended into it until it's full.
+ * Another thread could read from it at the same time (behind the writer).
+ */
+private[python] case class InMemoryRowQueue(page: MemoryBlock, fields: Int) extends RowQueue {
+  private val base: AnyRef = page.getBaseObject
+  private var last = page.getBaseOffset  // for writing
+  private var first = page.getBaseOffset  // for reading
```

Comment:
Maybe it would be clearer to name these `readOffset` and `writeOffset`? You could add a comment explaining that `writeOffset` points to, say, the first location where a new row would be written, while `readOffset` points to the start of the next row to read.